snuba/query/query.py

```diff
@@ -47,11 +48,42 @@ class Query:
     an abstract Snuba query and a concrete Clickhouse query, but
     that cannot come in this PR since it also requires a proper
     schema split in the dataset to happen.
+
+    NEW DATA MODEL:
+    The query is represented as a tree. The Query object is the root.
+    Nodes in the tree can have different types (Expression, Conditions, etc.)
+    Each node could be an individual node (like a column) or an collection.
+    A collection can be a sequence (like a list of Columns) or a hierarchy
+    (like function calls).
+
+    There are three ways to manipulate the query:
+    - traverse the tree. Traversing the full tree in an untyped way is
+      not extremely useful. What is more interesting is being able to
+      iterate over all the nodes of a given type (like expressions).
+      This is achieved through the NodeContainer interface.
+
+    - replace specific nodes. NodeContainer provides a map methods that
+      allows the callsite to apply a func to all nodes of a specific
+      type. This is useful for replacing expressions across the query.
+
+    - direct access to the root and explore specific parts of the tree
+      from there.
     """
     # TODO: Make getters non nullable when possible. This is a risky
     # change so we should take one field at a time.
 
-    def __init__(self, body: MutableMapping[str, Any], data_source: RelationalSource):
+    def __init__(self,
+        body: MutableMapping[str, Any],  # Temporary
+        data_source: RelationalSource,
+        # New data model to replace the one based on the dictionary
+        selected_columns: Optional[Sequence[Expression]] = None,
+        aggregations: Optional[Sequence[NodeAggregation]] = None,
+        array_join: Optional[Column] = None,
```

Comment:
Do you know if this is actually used anywhere? (If not, can we remove it?)

In my opinion, we should try to deprecate this long term and just replace it with a generic join keyword.

Reply:
I did not check, can verify. Actually removing it will be very convenient if possible.

Reply:
Unfortunately it is used for discoverV2. I need to dig deeper to understand where